### PR TITLE
benchmark: load-regions

### LINKED ISSUE #4994: Benchmark the `load-regions` at PD startup
## Enhancement Task

In the load regions stage when PD starts, in addition to reading, there will be delete operations. We need to benchmark the speed with N pieces of valid data and M pieces of invalid data.

In addition, we need to print the actual amount of data read and written in the log.

### PR DESCRIPTION
**Pull request by LLThomas**
Signed-off-by: LLThomas <zs033@qq.com>

### What problem does this PR solve?

Issue Number: Close #4994

### What is changed and how does it work?

```commit-message
This pull request add two benchmark function.
1. "BenchmarkLoadRegionsByVolume" will benchmark the LoadRegions() based on the number of regions (10000, 100000, 1000000).
2. "BenchmarkLoadRegionsByRandomMerge" will benchmark the LoadRegions() based on randomly merged regions.
```

### Check List

Tests

- Unit test

### Release note

```release-note
None.
```

### CODE REVIEW

**Comment by LLThomas** [comment]
Benchmark results :
```
goos: darwin
goarch: amd64
pkg: github.com/tikv/pd/server/storage
cpu: Intel(R) Core(TM) i5-6267U CPU @ 2.90GHz
BenchmarkLoadRegionsByVolume
BenchmarkLoadRegionsByVolume/input_size_10000
BenchmarkLoadRegionsByVolume/input_size_10000-4         	      26	  38594630 ns/op
BenchmarkLoadRegionsByVolume/input_size_100000
BenchmarkLoadRegionsByVolume/input_size_100000-4        	       2	 507323615 ns/op
BenchmarkLoadRegionsByVolume/input_size_1000000
BenchmarkLoadRegionsByVolume/input_size_1000000-4       	       1	6567605114 ns/op
PASS
```
```
goos: darwin
goarch: amd64
pkg: github.com/tikv/pd/server/storage
cpu: Intel(R) Core(TM) i5-6267U CPU @ 2.90GHz
BenchmarkLoadRegionsByRandomMerge
BenchmarkLoadRegionsByRandomMerge/random_merge
BenchmarkLoadRegionsByRandomMerge/random_merge-4         	       1	12396521818 ns/op
PASS
```

**Comment by nolouch** [comment]
```
BenchmarkLoadRegionsByVolume/input_size_1000000-4       	       1	6567605114 ns/op
BenchmarkLoadRegionsByRandomMerge/random_merge-4         	       1	12396521818 ns/op
```
They are a double relationship; if it load takes 10 minutes, it should store 100 million data? Can you help test it locally?

**Review by nolouch — COMMENTED** [review]
CI failed, seems do not make format.

**Comment by LLThomas** [comment]
> ```
> BenchmarkLoadRegionsByVolume/input_size_1000000-4       	       1	6567605114 ns/op
> BenchmarkLoadRegionsByRandomMerge/random_merge-4         	       1	12396521818 ns/op
> ```
> 
> They are a double relationship; if `load regions` takes `10 minutes`, it should store 100 million data? Can you help test it locally?

I tested it. "BenchmarkLoadRegionsByVolume(100000000)" ran for more than 15 minutes and did not stop, so I just ended it

**Comment by LLThomas** [comment]
The result of "BenchmarkLoadRegionsByVolume" (10000, 100000) are so strange. - -!
```
goos: darwin
goarch: amd64
pkg: github.com/tikv/pd/server/storage
cpu: Intel(R) Core(TM) i5-6267U CPU @ 2.90GHz
BenchmarkLoadRegionsByVolume
BenchmarkLoadRegionsByVolume/input_size_10000
    storage_test.go:411: region number is:  10000
    ...
BenchmarkLoadRegionsByVolume/input_size_10000-4         	1000000000	         0.03583 ns/op
BenchmarkLoadRegionsByVolume/input_size_100000
    storage_test.go:411: region number is:  100000
    ...
BenchmarkLoadRegionsByVolume/input_size_100000-4        	1000000000	         0.4880 ns/op
BenchmarkLoadRegionsByVolume/input_size_1000000
    storage_test.go:411: region number is:  1000000
BenchmarkLoadRegionsByVolume/input_size_1000000-4       	       1	6171993546 ns/op
PASS
```
```
goos: darwin
goarch: amd64
pkg: github.com/tikv/pd/server/storage
cpu: Intel(R) Core(TM) i5-6267U CPU @ 2.90GHz
BenchmarkLoadRegionsByRandomMerge
BenchmarkLoadRegionsByRandomMerge/merge_ratio_0
    storage_test.go:411: region number is:  1000000
BenchmarkLoadRegionsByRandomMerge/merge_ratio_0-4         	       1	6189519562 ns/op
BenchmarkLoadRegionsByRandomMerge/merge_ratio_20
    storage_test.go:411: region number is:  800000
BenchmarkLoadRegionsByRandomMerge/merge_ratio_20-4        	       1	9092247731 ns/op
BenchmarkLoadRegionsByRandomMerge/merge_ratio_40
    storage_test.go:411: region number is:  600000
BenchmarkLoadRegionsByRandomMerge/merge_ratio_40-4        	       1	11706250144 ns/op
BenchmarkLoadRegionsByRandomMerge/merge_ratio_60
    storage_test.go:411: region number is:  400000
BenchmarkLoadRegionsByRandomMerge/merge_ratio_60-4        	       1	14067181622 ns/op
BenchmarkLoadRegionsByRandomMerge/merge_ratio_80
    storage_test.go:411: region number is:  200000
BenchmarkLoadRegionsByRandomMerge/merge_ratio_80-4        	       1	17723740556 ns/op
PASS
```

File: server/storage/storage_test.go [hunk 1]
```diff
@@ -279,3 +281,169 @@ func TestLoadRegionsExceedRangeLimit(t *testing.T) {
...
+	defer func() {
+		err = lb.Close()
+		if err != nil {
+			b.Fatal(err)
+		}
+	}()
+
+	b.ResetTimer()
+	err = lb.LoadRegions(ctx, cluster.CheckAndPutRegion)
+	//b.Log("region number is: ", cluster.GetRegionCount())
```
Comment (nolouch) [inline server/storage/storage_test.go:411]:
```suggestion
	// b.Log("region number is: ", cluster.GetRegionCount())
```

**Comment by nolouch** [comment]
plz check the ci fail reason: https://github.com/tikv/pd/runs/6978495329?check_suite_focus=true

**Comment by LLThomas** [comment]
> plz check the ci fail reason: https://github.com/tikv/pd/runs/6978495329?check_suite_focus=true

Okk, I forgot about this error.😭

**Review by nolouch — APPROVED** [review]
lgtm

File: server/storage/storage_test.go [hunk 2]
```diff
@@ -279,3 +281,166 @@ func TestLoadRegionsExceedRangeLimit(t *testing.T) {
...
+		err := lb.SaveRegion(region)
+		if err != nil {
+			return err
+		}
+	}
+
+	return lb.Flush()
+}
+
+func benchmarkLoadRegions(n int, b *testing.B, ratio int) {
```
Comment (HunDunDM) [inline server/storage/storage_test.go:387]: Better to put `b *testing.B` first.

File: server/storage/storage_test.go [hunk 3]
```diff
@@ -279,3 +281,166 @@ func TestLoadRegionsExceedRangeLimit(t *testing.T) {
...
+			regions[mergeIndex].StartKey = regions[pos].StartKey
+		} else {
+			for {
+				mergeIndex--
+				if _, ok := note[mergeIndex]; !ok {
+					break
+				}
+			}
+			if mergeIndex != -1 {
+				regions[pos].StartKey = regions[mergeIndex].StartKey
```
Comment (HunDunDM) [inline server/storage/storage_test.go:338]:
```suggestion
				regions[pos].EndKey = regions[mergeIndex].EndKey
```
It looks like it should be the end key here?

Comment (LLThomas) [reply server/storage/storage_test.go:338]: When we find a "mergeIndex" before "pos", I think we should reset the pos region StartKey to the mergeIndex region StartKey for the purpose of merging mergeIndex region to pos region.
```
Origin:

Region(mergeIndex)                             Region(pos)
|-------------------------------------------|  |--------------------------------------------|
mergeIndex(StartKey)       mergeIndex(EndKey)  pos(StartKey)                      pos(EndKey)


Reset:

Region(mergeIndex)
|-------------------------------------------| 
mergeIndex(StartKey)       mergeIndex(EndKey)
Region(pos)
|------------------------------------------------------------------------------------------|
mergeIndex(StartKey)                                                             pos(EndKey)
```

And when I set "egions[pos].EndKey = regions[mergeIndex].EndKey", the benchmark log is (input is 100):
```
goos: darwin
goarch: amd64
pkg: github.com/tikv/pd/server/storage
cpu: Intel(R) Core(TM) i5-6267U CPU @ 2.90GHz
BenchmarkLoadRegionsByRandomMerge/merge_ratio_0-4         	1000000000	         0.001031 ns/op	       0 B/op	       0 allocs/op
--- BENCH: BenchmarkLoadRegionsByRandomMerge/merge_ratio_0-4
    storage_test.go:408: region number is:  100
    ...
BenchmarkLoadRegionsByRandomMerge/merge_ratio_20-4        	1000000000	         0.0007170 ns/op	       0 B/op	       0 allocs/op
--- BENCH: BenchmarkLoadRegionsByRandomMerge/merge_ratio_20-4
    storage_test.go:408: region number is:  100
    ...
BenchmarkLoadRegionsByRandomMerge/merge_ratio_40-4        	1000000000	         0.0009663 ns/op	       0 B/op	       0 allocs/op
--- BENCH: BenchmarkLoadRegionsByRandomMerge/merge_ratio_40-4
    storage_test.go:408: region number is:  100
    ...
BenchmarkLoadRegionsByRandomMerge/merge_ratio_60-4        	1000000000	         0.001285 ns/op	       0 B/op	       0 allocs/op
--- BENCH: BenchmarkLoadRegionsByRandomMerge/merge_ratio_60-4
    storage_test.go:408: region number is:  96
    ...
BenchmarkLoadRegionsByRandomMerge/merge_ratio_80-4        	1000000000	         0.0007771 ns/op	       0 B/op	       0 allocs/op
--- BENCH: BenchmarkLoadRegionsByRandomMerge/merge_ratio_80-4
    storage_test.go:408: region number is:  96
    ...
PASS
```
The region number is not correct. For example, when ration is 20, the region number should be 80

**Comment by LLThomas** [comment]
The benchmark result:
```
goos: darwin
goarch: amd64
pkg: github.com/tikv/pd/server/storage
cpu: Intel(R) Core(TM) i5-6267U CPU @ 2.90GHz
BenchmarkLoadRegionsByVolume
BenchmarkLoadRegionsByVolume/input_size_10000
    storage_test.go:388: region number is:  10000
    ...
BenchmarkLoadRegionsByVolume/input_size_10000-4         	1000000000	         0.05243 ns/op
BenchmarkLoadRegionsByVolume/input_size_100000
    storage_test.go:388: region number is:  100000
    ...
BenchmarkLoadRegionsByVolume/input_size_100000-4        	1000000000	         0.4940 ns/op
BenchmarkLoadRegionsByVolume/input_size_1000000
    storage_test.go:388: region number is:  1000000
BenchmarkLoadRegionsByVolume/input_size_1000000-4       	       1	6164669892 ns/op
PASS
```

```
goos: darwin
goarch: amd64
pkg: github.com/tikv/pd/server/storage
cpu: Intel(R) Core(TM) i5-6267U CPU @ 2.90GHz
BenchmarkLoadRegionsByRandomMerge
BenchmarkLoadRegionsByRandomMerge/merge_ratio_0
    storage_test.go:388: region number is:  1000000
BenchmarkLoadRegionsByRandomMerge/merge_ratio_0-4         	       1	6901512368 ns/op
BenchmarkLoadRegionsByRandomMerge/merge_ratio_20
    storage_test.go:388: region number is:  800000
BenchmarkLoadRegionsByRandomMerge/merge_ratio_20-4        	       1	10908678807 ns/op
BenchmarkLoadRegionsByRandomMerge/merge_ratio_40
    storage_test.go:388: region number is:  600000
BenchmarkLoadRegionsByRandomMerge/merge_ratio_40-4        	       1	11938138820 ns/op
BenchmarkLoadRegionsByRandomMerge/merge_ratio_60
    storage_test.go:388: region number is:  400000
BenchmarkLoadRegionsByRandomMerge/merge_ratio_60-4        	       1	13942839723 ns/op
BenchmarkLoadRegionsByRandomMerge/merge_ratio_80
    storage_test.go:388: region number is:  200000
BenchmarkLoadRegionsByRandomMerge/merge_ratio_80-4        	       1	16720773570 ns/op
PASS
```

**Comment by LLThomas** [comment]
func randomMerge(regions []*metapb.Region, n int, ratio int) {
	rand.Seed(6)
	note := make(map[int]bool)
	for i := 0; i < n*ratio/100; i++ {
		pos := rand.Intn(n)
		for {
			if _, ok := note[pos]; !ok {
				break
			}
			pos = rand.Intn(n)
		}
		note[pos] = true

		mergeIndex := pos + 1
		for mergeIndex < n {
			_, ok := note[mergeIndex]
			if ok {
				mergeIndex++
			} else {
				break
			}
		}
		if mergeIndex < n {
			regions[mergeIndex].StartKey = regions[pos].StartKey
		}
		if mergeIndex == n {
			if pos == n-1 {
				preMergeIndex := pos - 1
				for preMergeIndex >= 0 {
					_, ok := note[preMergeIndex]
					if ok {
						preMergeIndex--
					} else {
						break
					}
				}
				regions[mergeIndex-1].StartKey = regions[preMergeIndex].StartKey
				note[preMergeIndex] = true
			} else {
				regions[mergeIndex-1].StartKey = regions[pos].StartKey
			}
		}
	}
}

**Comment by LLThomas** [comment]
```
goos: darwin
goarch: amd64
pkg: github.com/tikv/pd/server/storage
cpu: Intel(R) Core(TM) i5-6267U CPU @ 2.90GHz
BenchmarkLoadRegionsByRandomMerge
BenchmarkLoadRegionsByRandomMerge/merge_ratio_0
    storage_test.go:400: region number is:  1000000
BenchmarkLoadRegionsByRandomMerge/merge_ratio_0-4         	       1	6914356195 ns/op
BenchmarkLoadRegionsByRandomMerge/merge_ratio_20
    storage_test.go:400: region number is:  800000
BenchmarkLoadRegionsByRandomMerge/merge_ratio_20-4        	       1	9428809801 ns/op
BenchmarkLoadRegionsByRandomMerge/merge_ratio_40
    storage_test.go:400: region number is:  600000
BenchmarkLoadRegionsByRandomMerge/merge_ratio_40-4        	       1	11928843631 ns/op
BenchmarkLoadRegionsByRandomMerge/merge_ratio_60
    storage_test.go:400: region number is:  400000
BenchmarkLoadRegionsByRandomMerge/merge_ratio_60-4        	       1	14017329022 ns/op
BenchmarkLoadRegionsByRandomMerge/merge_ratio_80
    storage_test.go:400: region number is:  200000
BenchmarkLoadRegionsByRandomMerge/merge_ratio_80-4        	       1	16545684329 ns/op
PASS
```

**Review by HunDunDM — APPROVED** [review]

**Pull request merged**